[#20] 유저, 로그인 로직 수정

daakludens opened the pull request:
유닛 테스트 작성 도중 발견한 로직 에러 수정 및 추가, 삭제

daakludens committed:
[#20] fix : 유저와 로그인 로직 수정

daakludens committed:
[#20] fix : 유저 정보 가져오기와 로그인 기능 분리

daakludens committed:
[#20] fix : UserService 로그인 로직 수정

daakludens committed:
[#20] fix : updateUserInfo 중복적인 파라미터 제거

f-lab-dev commented on src/main/java/com/ludensdomain/controller/UserController.java:
```diff
@@ -37,17 +38,9 @@
-        }
-
-        return result;
+        return (user == null) ? RESPONSE_BAD_REQUEST : RESPONSE_OK;
```
```suggestion
        return user == null ? RESPONSE_BAD_REQUEST : RESPONSE_OK;
```

daakludens replied:
아.. 불필요한 괄호 계속 까먹네요.. 다음에 꼭 기억하고 사용하겠습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/controller/UserController.java:
```diff
@@ -67,24 +60,33 @@
-    @PutMapping("{id}")
-    public void update(@RequestBody @Valid UserDto user, @PathVariable long id) {
+    @LoginCheck
+    @PutMapping("update")
```
여기는 중복표현이 들어있네요. 어떤 점 때문일까요?

daakludens replied:
이전에 파라미터로 user 정보와 id를 따로 가져왔었는데 user에 id 정보가 이미 담겨 있으니 중복적인 파라미터라 판단했습니다. 그래서 @PathVariable로 받아오는 "{id}" 부분을 기능을 나타내는 "update"로 변경했습니다. 
아예 url 명을 빼고 진행했습니다. 추가로 사용자 추가 로직(signUp)도 중복적인 표현인 거 같아 지웠습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/mapper/UserMapper.java:
```diff
@@ -10,7 +10,11 @@
+
+    void changePassword(long id, String password);
+
+    Boolean checkIdExists(long id);
```
여기는 null을 리턴할 가능성이 있나요? 없다면 primitive type을 쓰는게 낫지 않을까요?

daakludens replied:
무조건 true 혹은 false를 반환하기에 primitive type인 boolean을 사용했습니다.
Boolean보다 boolean이 메모리 사용량에서 훨씬 적고 효율적임으로 결과가 보장된다면 boolean으로 사용하는 게 좋습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/service/UserService.java:
```diff
@@ -1,39 +1,46 @@
+    @Transactional
+    public UserDto login(long id, String password) {
+        UserDto existedUser = Optional.ofNullable(getUserInfo(id)).orElseThrow(NonExistingUserException::new);
+        String encryptPw = stringEncryptor.encrypt(password);
```
개행은 적절해 해줘야 가독성이 늘어납니다~

daakludens replied:
네, 수정하겠습니다.

f-lab-dev commented on src/main/java/com/ludensdomain/service/UserService.java:
```diff
@@ -1,39 +1,46 @@
+        String encryptPw = stringEncryptor.encrypt(password);
+        if (encryptPw.equals(existedUser.getPassword())) {
+            loginService.login(id, existedUser.getRole());
+            return existedUser;
```
굳이 로그인의 결과로 유저의 정보를 리턴해줄 필요가 있을까요?

daakludens replied:
로그인 후 유저 내용이 바로 나올 수 있도록 로직이 필요하다 생각했는데... 지금 보니 컨트롤러에선 반환된 유저 정보를 단순히 정보 유무로만 확인하니 불필요한 정보입니다.
그래서 boolean을 반환해 ResponseEntity를 정의하도록 설정하고, 유저 정보를 가져오는 메서드를 따로 정의했습니다.

daakludens committed:
[#20] fix : 첨삭한 내용 적용
- Boolean을 boolean으로 변경
- 로그인 기능에서 user 정보 반환을 로그인 성공 여부로 변경
- url 이름 수정
- UserService의 login 메서드 가독성을 위해 개행

f-lab-dev commented on src/main/java/com/ludensdomain/service/UserService.java:
```diff
@@ -20,14 +20,17 @@
+            return true;
         } else {
-            return null;
+            return false;
```
로그인에 실패하면 예외를 던지는게 나을 것 같습니다.
연산의 결과로 boolean이나 기타 값들을 리턴하는것은 정상적인 흐름으로 처리하는것이기에 비정상적인 흐름임을 알려주는 예외를 사용하는게 낫습니다

daakludens replied:
아, 그러네요. 여부가 아닌 일치/불일치를 다루는 로직이니 예외를 발생시켜 문제가 발생했다는 걸 보여주는 게 도 좋은 흐름이겠네요. 수정하겠습니다.

daakludens committed:
[#20] fix : 서비스 레이어에서 로그인 패스워드 불일치 시 예외를 던지도록 수정, 패스워드 불일치 예외 unchecked exception 추가
